components/__tests__/ContainerMembershipLoginWithEmail.spec.js

```diff
@@ -0,0 +1,144 @@
+import flushPromises from 'flush-promises'
+import ContainerMembershipLoginWithEmail from '../ContainerMembershipLoginWithEmail.vue'
+import createWrapperHelper from '~/test/helpers/createWrapperHelper'
+import UIMembershipEmailInput from '~/components/UIMembershipEmailInput.vue'
+
+const createWrapper = createWrapperHelper({
+  mocks: {
+    $fire: {
+      authReady: jest.fn(),
+    },
+  },
+})
+
+describe('handle event emitted by child component UIMembershipEmailInput', function () {
+  test('handle "input" event emitted by UIMembershipEmailInput', function () {
+    const wrapper = createWrapper(ContainerMembershipLoginWithEmail)
+    const emailInput = wrapper.findComponent(UIMembershipEmailInput)
+    const mockInputValue = 'mockInputValue'
+    emailInput.vm.$emit('input', mockInputValue)
+    expect(wrapper.vm.emailInput).toBe(mockInputValue)
+  })
+  test('handle "inputValidStateChange" event emitted by UIMembershipEmailInput', function () {
+    const wrapper = createWrapper(ContainerMembershipLoginWithEmail)
+    const emailInput = wrapper.findComponent(UIMembershipEmailInput)
+    const mockEmailInputValidState = true
+    emailInput.vm.$emit('inputValidStateChange', mockEmailInputValidState)
+    expect(wrapper.vm.isEmailInputValid).toBe(mockEmailInputValidState)
+  })
+})
+
+describe('@nuxtjs/firebase service', function () {
+  test('should lazyload @nuxtjs/firebase auth service on component created', async function () {
+    expect.assertions(1)
```

Comment:
加這個有什麼好處嗎？我看 [Jest 官方文檔](https://jestjs.io/docs/en/expect#expectassertionsnumber)是說，它可以確保非同步的測試有被確實呼叫。但這裡以及其它測試，似乎都跟這個沒什麼關係？

Reply:
有一個慣例是
> 永遠要在測試開頭用 expect.assertions(number) 來確保有多少 assertions 會被執行，免得在非同步的情境下有 assertions 不小心被跳過了，造成 false positives。
不過確實我有點把它當成習慣而永遠會加 `expect.assertions(number)`，但思考了一下非同步程式應該也只有 promise or callback 的寫法可能有被跳過的疑慮，async await 似乎不太會發生。

Reply:
是有 `async/await`，但它不會導致 `expect` 被重複呼叫或不被呼叫，Jest 的官方範例是有這個可能的，因此加這個確實有必要。

非必要地加這個也沒什麼缺點，能想到的只有未來多加或少加一個 `expect` 時要改這一段。優點嘛，能想到的只有避免有必要時忘記加這一段。

非必要情境下有沒有需要加這一段，我會希望它能變成一個共識。以免後來的人搞不清楚為何有些加有些不加。我目前是傾向於沒必要時不加。可以知道你引用的出處嗎？

Reply:
從 [Testing Vue.js Applications](https://www.manning.com/books/testing-vue-js-applications) 一書中參考的，
其中 4.5.2. Testing asynchronous code 一節提到：
> The reason you should set the number of assertions in an asynchronous test is to make sure all the assertions execute before the test finishes.

```js
test('fetches data', async () => {
  expect.assertions(1)
  const data = await fetchListData()
  expect(data).toBe('some data')
})
```

> Asynchronous tests are the biggest cause of false positives. Without the Jest expect.assertions call, if an assertion is inside an asynchronous action but the test doesn’t know that it’s asynchronous, the test will pass because it never executes the assertion.

```js
test('awaits promise', async () => {
  expect.assertions(1)
  let hasResolved = false
  Promise.resolve().then(() => {
    hasResolved = true
  })

  expect(hasResolved).toBe(true)
})
```

Reply:
ok，那就之後只要測試中牽涉到非同步的都要加 `expect.assertions`，但沒用到的就不加囉？

Reply:
> ok，那就之後只要測試中牽涉到非同步的都要加 `expect.assertions`，但沒用到的就不加囉？

同意